subprojects/core-api/src/main/java/org/gradle/api/provider/FileContents.java

```diff
@@ -0,0 +1,42 @@
+/*
+ * Copyright 2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.provider;
+
+import org.gradle.api.Incubating;
+
+/**
+ * Provides lazy access to the contents of a given file.
+ *
+ * @since 6.1
+ */
+@Incubating
+public interface FileContents {
+
+    /**
+     * Gets a provider to the entire file contents as a single String.
+     *
+     * @return provider to the entire file contents as a single String.
+     */
+    Provider<String> getAsText();
```

Comment:
I would add these methods directly to `RegularFile`, so that a `RegularFile` can provide content in addition to information about the file. Or, if we want to keep these separate, `FileContents` should probably go and live in the same packages as the other file types.

We should also consider how this relates to `TextResource`. Possibly we replace `TextResource` with these new methods and merge some of the conveniences into `RegularFile` or `FileContents`.

Also, we should clarify in the docs what happens when the referenced thing cannot be read for whatever reason, or does not exist, or is not defined.

Reply:
In the end I preferred to keep the methods separate from `RegularFile` to avoid introducing a dependency on providers from there and to keep all the _build logic inputs_ API in a single place as it matures.

I looked briefly at `TextResource` as I was implementing this and my conclusion was that it was better to introduce an independent facility (`FileContents`) with the minimal required context.

Thanks for the review, @adam. I will:
- [ ] move `FileContents` together with the other file types
- [ ] clarify those conditions in the docs

Reply:
Done.